src/google/protobuf/compiler/csharp/csharp_generator.cc

```diff
@@ -71,15 +79,31 @@ bool Generator::Generate(
     return false;
   }
 
-  std::string file_extension = ".cs";
-  std::string base_namespace = "";
-  bool generate_directories = false;
+  // -----------------------------------------------------------------
+  // parse generator options
+  //
+  // The following options are supported:
+  // 1. file_extension: Defaults to ".cs"
+  // 2. base_namespace: Defaults to ""
+  // 3. generate_directories: Defaults to false
+  // 4. access_level: Defaults to "public"
+
+  struct Options file_options;
+
+  //std::string file_extension = ".cs";
```

Comment:
Please remove the commented out lines. It sounds like really the Options struct should know how to parse a `vector<pair<string, string>>`, rather than the code being here.

As noted before, `file_options` should be `generator_options` or similar.